clusterloader2/pkg/measurement/common/network/network_performance_measurement.go

```diff
@@ -0,0 +1,493 @@
+}
+
+func (npm *networkPerformanceMeasurement) calculateMetricDataValue(dataElem *measurementutil.DataItem, metricIndex int) {
+	var aggregatePodPairMetrics []float64
+	var metricResponse []float64
+
+	npm.metricLock.Lock()
+	defer npm.metricLock.Unlock()
+
+	switch npm.podRatioType {
+	case oneToOne:
+		for _, metricResponse = range npm.metricVal {
+			if len(metricResponse) > 0 {
+				dataElem.Data[value] = metricResponse[metricIndex]
+			} else {
+				dataElem.Data[value] = 0
+			}
+		}
+		klog.Info("Metric value: ", dataElem.Data[value])
+	case manyToMany:
+		for _, metricResponse = range npm.metricVal {
+			if len(metricResponse) > 0 {
+				aggregatePodPairMetrics = append(aggregatePodPairMetrics, metricResponse[metricIndex])
+			}
+		}
+		dataElem.Data[perc05] = getPercentile(aggregatePodPairMetrics, percentile05)
+		dataElem.Data[perc50] = getPercentile(aggregatePodPairMetrics, percentile50)
+		dataElem.Data[perc95] = getPercentile(aggregatePodPairMetrics, percentile95)
+		klog.Info("Aggregate Metric value: ", aggregatePodPairMetrics)
+	}
+	klog.Infof("PodPairNumber: %v, Response received number: %v ", npm.podPairNumber, len(npm.metricVal))
+}
+
+func (npm *networkPerformanceMeasurement) createResultSummary() {
+	var resultSummary testResultSummary
+	switch npm.protocol {
+	case protocolTCP:
+		npm.getMetricData(&resultSummary, tcpBandwidth, throughput)
+		resultSummary.protocol = protocolTCP
+	case protocolUDP:
+		npm.getMetricData(&resultSummary, udpPacketPerSecond, packetPerSecond)
+		npm.getMetricData(&resultSummary, udpJitter, jitter)
+		npm.getMetricData(&resultSummary, udpLatencyAverage, latency)
+		resultSummary.protocol = protocolUDP
+	case protocolHTTP:
+		npm.getMetricData(&resultSummary, httpResponseTime, responseTime)
+		resultSummary.protocol = protocolHTTP
+	}
+	resultSummary.service = "P2P"
+	resultSummary.podRatio = npm.podRatioType
+	npm.resultSummary = resultSummary
+}
+
+func (npm *networkPerformanceMeasurement) getMetricData(data *testResultSummary, metricIndex int, metricName string) {
+	var metricDataItem measurementutil.DataItem
+	metricDataItem.Data = make(map[string]float64)
+	metricDataItem.Labels = make(map[string]string)
+	metricDataItem.Labels["Metric"] = metricName
+	npm.calculateMetricDataValue(&metricDataItem, metricIndex)
+	metricDataItem.Unit = metricUnitMap[metricDataItem.Labels["Metric"]]
+	data.DataItems = append(data.DataItems, metricDataItem)
+	klog.V(3).Infof("TestResultSummary: %v", data)
+}
+
+type podList []workerPodData
+type workerPodInfo struct {
+	podPairCh    chan uniquePodPair
+	workerPodMap map[string]podList
+}
+
+func (l *podList) Pop() (*workerPodData, error) {
+	if len(*l) == 0 {
+		return nil, fmt.Errorf("empty pod list")
+	}
+	lastElement := (*l)[len(*l)-1]
+	*l = (*l)[:len(*l)-1]
+	return &lastElement, nil
+}
+
+func (wp *workerPodInfo) formUniquePodPair() {
+	for index := 0; len(wp.workerPodMap) > 0; index++ {
```

Comment:
> Ya even i am wondering , how do we solve this problem of 2 pods on same node remaining without pairing. How do we know the longest queue in map. We can implement it , but a costly one in terms of performance . Everytime , we may have to iterate and figure out the longest queue.

How many pods you want to support? For 100 pods it doesn't really matter.

Anyway - I actually think it's relatively easy to implement here, because the greedy approach works here.
You always pick:
- first pod from the node with the biggest number of pods
- second pod from the second node
So it's basically enough to keep heap with (node, #pods on that node) sorted by #pods:
https://golang.org/pkg/container/heap/

And the algorithm is basically:
while not empty {
  first := heap.Pop()
  second := heap.Pop()
  if first.numberOfPods > 1 {
    heap.Push({first.node, first-1})
  }
  // same for second
  uniquePodPair = (first, second)
}

Reply:
modified